src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -505,7 +503,31 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
         path_str: &'path str,
         ns: Namespace,
         module_id: DefId,
-        extra_fragment: &Option<UrlFragment>,
+        user_fragment: &Option<UrlFragment>,
+    ) -> Result<(Res, Option<UrlFragment>), ErrorKind<'path>> {
+        let (res, rustdoc_fragment) = self.resolve_inner(path_str, ns, module_id)?;
+        let chosen_fragment = match (user_fragment, rustdoc_fragment) {
+            (Some(_), Some(r_frag)) => {
+                let diag_res = match r_frag {
+                    UrlFragment::Def(_, did) => Res::Def(self.cx.tcx.def_kind(did), did),
+                    // FIXME: eliminate this branch somehow
```

Comment:
we can pass in `user_fragment: Option<(FragmentKind, DefId)>`

Reply:
Actually there'd be more changes involved. I'll do that now.